component/src/main/java/io/siddhi/extension/io/kafka/metrics/SinkMetrics.java

```diff
@@ -0,0 +1,145 @@
+        return MetricsDataHolder.getInstance().getMetricService()
+                .counter(String.format("io.siddhi.SiddhiApps.%s.Siddhi.Kafka.Sink.Errors.Per.Stream.%s.%s.%s.%s",
+                        siddhiAppName, topic, "stream_id." + streamId, "partition." + partition, "errorString." +
+                                errorString), Level.INFO);
+    }
+
+    public void getLastMessageSize(String topic, int partition, String streamId, double messageSize) {
+        updateMessageSizeMap(topic, partition, messageSize);
+        MetricsDataHolder.getInstance().getMetricService()
+                .gauge(String.format("io.siddhi.SiddhiApps.%s.Siddhi.Kafka.Sink.Per.Stream.%s.%s.%s.%s",
+                        siddhiAppName, topic, "partition." + partition,
+                        "streamId." + streamId, "last_message_size_in_bytes"),
+                        Level.INFO, () -> messageSizeMap.get(topic).get(partition));
+    }
+
+    public void getLastMessageAckLatency(String topic, int partition, String streamId, long latency) {
+        updateLatencyMap(topic, partition, latency);
+        MetricsDataHolder.getInstance().getMetricService()
+                .gauge(String.format("io.siddhi.SiddhiApps.%s.Siddhi.Kafka.Sink.Per.Stream.%s.%s.%s.%s",
+                        siddhiAppName, topic, "partition." + partition,
+                        "streamId." + streamId, "last_message_latency_in_millis"),
+                        Level.INFO, () -> latencyMap.get(topic).get(partition));
+    }
+
+    public void getLastCommittedOffset(String topic, int partition, String streamId, long offset) {
+        updateOffsetMap(topic, partition, offset);
+        MetricsDataHolder.getInstance().getMetricService()
+                    .gauge(String.format("io.siddhi.SiddhiApps.%s.Siddhi.Kafka.Sink.Current.Offset.%s.%s.%s",
+                            siddhiAppName, topic, "partition." + partition, "streamId." + streamId), Level.INFO,
+                            () -> offsetMap.get(topic).get(partition));
+    }
+
+    public void getLastMessagePublishedTime(String topic, int partition, String streamId, long pushedTimestamp) {
+        updateLastMessagePublishedTimeMap(topic, partition, pushedTimestamp);
+        MetricsDataHolder.getInstance().getMetricService()
+                .gauge(String.format("io.siddhi.SiddhiApps.%s.Siddhi.Kafka.Sink.Per.Stream.%s.%s.%s.%s",
+                        siddhiAppName, topic, "partition." + partition,
+                        "streamId." + streamId, "last_message_published_at"),
+                        Level.INFO, System::currentTimeMillis);
+    }
+
+    private void updateOffsetMap(String topic, int partition, long offset) {
+        Map<Integer, Long> partitionMap;
+        if (offsetMap.get(topic) == null) {
+            partitionMap = new ConcurrentHashMap();
+        } else {
+            partitionMap = offsetMap.get(topic);
+        }
```

Comment:
Is `if` block executed only once, and `else` block rest of the time?
If so, shall we swap these conditions?

Reply:
Could you please elaborate more on this? I think I did not get your point.

Reply:
Let's say the following `if` block executes only once:
```java
if (offsetMap.get(topic) == null) {
        partitionMap = new ConcurrentHashMap();
}
```

and thereafter, the `else` block is going to be executed, every time.

If that's the case, we can swap the conditions like following:
```java
        if (offsetMap.get(topic) != null) {
            partitionMap = offsetMap.get(topic);
        } else {
            partitionMap = new ConcurrentHashMap();
        }
```

Reply:
From what I understand, whatever the condition that we put on the if condition will be evaluated. So it's either `offsetMap.get(topic) == null` or `offsetMap.get(topic) != null`
and based on the outcome, one of the following would be exceuted:
- `partitionMap = new ConcurrentHashMap();`
- `partitionMap = offsetMap.get(topic);`

So either way, one condition and one statement will be executed. I do not see a value addition of changing the if-else block as being suggested.